

<!doctype html>
<html lang="en" class="no-js">
  <head>
    

<meta charset="utf-8">



<!-- begin SEO -->









<title>Posts by Collection - Home</title>







<meta property="og:locale" content="en-US">
<meta property="og:site_name" content="Home">
<meta property="og:title" content="Posts by Collection">


  <link rel="canonical" href="http://localhost:4000/collection-archive/">
  <meta property="og:url" content="http://localhost:4000/collection-archive/">







  

  












  <script type="application/ld+json">
    {
      "@context" : "http://schema.org",
      "@type" : "Person",
      "name" : "Your Name",
      "url" : "http://localhost:4000",
      "sameAs" : null
    }
  </script>






<!-- end SEO -->


<link href="http://localhost:4000/feed.xml" type="application/atom+xml" rel="alternate" title="Home Feed">

<!-- http://t.co/dKP3o1e -->
<meta name="HandheldFriendly" content="True">
<meta name="MobileOptimized" content="320">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<script>
  document.documentElement.className = document.documentElement.className.replace(/\bno-js\b/g, '') + ' js ';
</script>

<!-- For all browsers -->
<link rel="stylesheet" href="http://localhost:4000/assets/css/main.css">

<meta http-equiv="cleartype" content="on">
    

<!-- start custom head snippets -->

<!-- favicon from https://commons.wikimedia.org/wiki/File:OOjs_UI_icon_academic-progressive.svg -->
<link rel="apple-touch-icon" sizes="180x180" href="http://localhost:4000/images/apple-touch-icon-180x180.png"/>
<link rel="icon" type="image/svg+xml" href="http://localhost:4000/images/favicon.svg"/>
<link rel="icon" type="image/png" href="http://localhost:4000/images/favicon-32x32.png" sizes="32x32"/>
<link rel="icon" type="image/png" href="http://localhost:4000/images/favicon-192x192.png" sizes="192x192"/>
<link rel="manifest" href="http://localhost:4000/images/manifest.json"/>
<link rel="icon" href="/images/favicon.ico"/>
<meta name="theme-color" content="#ffffff"/>
<link rel="stylesheet" href="http://localhost:4000/assets/css/academicons.css"/>


<!-- Support for MatJax -->
<script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
<script id="MathJax-script" async src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>

<!-- end custom head snippets -->

  </head>

  <body>

    <!--[if lt IE 9]>
<div class="notice--danger align-center" style="margin: 0;">You are using an <strong>outdated</strong> browser. Please <a href="http://browsehappy.com/">upgrade your browser</a> to improve your experience.</div>
<![endif]-->
    

<div class="masthead">
  <div class="masthead__inner-wrap">
    <div class="masthead__menu">
      <nav id="site-nav" class="greedy-nav">
        <button><div class="navicon"></div></button>
        <ul class="visible-links">
          <li class="masthead__menu-item masthead__menu-item--lg persist"><a href="http://localhost:4000/">Home</a></li>
          
            
            <li class="masthead__menu-item"><a href="http://localhost:4000/publications/">Publications</a></li>
          
            
            <li class="masthead__menu-item"><a href="http://localhost:4000/talks/">Talks</a></li>
          
            
            <li class="masthead__menu-item"><a href="http://localhost:4000/year-archive/">Blog Posts</a></li>
          
            
            <li class="masthead__menu-item"><a href="http://localhost:4000/cv/">CV</a></li>
          
            
            <li class="masthead__menu-item"><a href="http://localhost:4000/softwares/">Softwares</a></li>
          
            
            <li class="masthead__menu-item"><a href="http://localhost:4000/community_involvement/">Community Involvement</a></li>
          
          <li id="theme-toggle" class="masthead__menu-item persist tail">
            <a><i id="theme-icon" class="fa-solid fa-sun" aria-hidden="true" title="toggle theme"></i></a>
          </li>
        </ul>
        <ul class="hidden-links hidden"></ul>
      </nav>
    </div>
  </div>
</div>

    



<div id="main" role="main">
  


  <div class="sidebar sticky">
  



<div itemscope itemtype="http://schema.org/Person">

  <div class="author__avatar">
    
    	<img src="http://localhost:4000/images/Logo.png" class="author__avatar" alt="Sébastien Gradit, PhD">
    
  </div>

  <div class="author__content">
    <h3 class="author__name">Sébastien Gradit, PhD</h3>
    
    <p class="author__bio">Bioinformatician leveraging AI to advance diagnostics and therapeutics using genomic and multi-omics data.</p>
  </div>

  <div class="author__urls-wrapper">
    <button class="btn btn--inverse">Follow</button>
    <ul class="author__urls social-icons">
      <!-- Font Awesome icons / Biographic information  -->
      
        <li class="author__desktop"><i class="fa-solid fa-location-dot icon-pad-right" aria-hidden="true"></i>Paris</li>
      
      
        <li class="author__desktop"><i class="fas fa-fw fa-building-columns icon-pad-right" aria-hidden="true"></i>Institut Pasteur</li>
      
      
      
        <li><a href="mailto:sebastiengradit@gmail.com"><i class="fas fa-fw fa-envelope icon-pad-right" aria-hidden="true"></i>Email</a></li>
      

      <!-- Font Awesome and Academicons icons / Academic websites -->
            
      
        <li><a href="https://scholar.google.com/citations?view_op=list_works&hl=fr&user=1676I_AAAAAJ"><i class="ai ai-google-scholar icon-pad-right"></i>Google Scholar</a></li>
      
      
      
      
        <li><a href=" https://orcid.org/0000-0003-2690-8503"><i class="ai ai-orcid ai-fw icon-pad-right"></i>ORCID</a></li>
      
                              
      
      

      <!-- Font Awesome icons / Repositories and software development -->
      
            
            
      
        <li><a href="https://github.com/sebgra"><i class="fab fa-fw fa-github icon-pad-right" aria-hidden="true"></i>GitHub</a></li>
      
      
        <li><a href="https://kaggle.com/sbastiengradit"><i class="fab fa-fw fa-kaggle icon-pad-right" aria-hidden="true"></i>Kaggle</a></li>
            
      
        <li><a href="https://www.stackoverflow.com/users/10441276"><i class="fab fa-fw fa-stack-overflow icon-pad-right" aria-hidden="true"></i>Stackoverflow</a></li>
            

      <!-- Font Awesome icons / Social media -->
      
      
            
      
                  
                  
      
            
            
      
        <li><a href="https://www.linkedin.com/in/sebastien-gradit"><i class="fab fa-fw fa-linkedin icon-pad-right" aria-hidden="true"></i>LinkedIn</a></li>
            
      
        <li><a href="https://genomic.social/@sebgra"><i class="fab fa-fw fa-mastodon icon-pad-right" aria-hidden="true"></i>Mastodon</a></li>
      
            
                  
            
      
            
            
      
        <li><a href="https://twitter.com/segradit"><i class="fab fa-fw fa-x-twitter icon-pad-right" aria-hidden="true"></i>X (formerly Twitter)</a></li>
      
              
      
                      
      
      
            
    </ul>
  </div>
</div>

  
  </div>


  <div class="archive">
    
      <h1 class="page__title">Posts by Collection</h1>
    
    



  
  
    
  
    
  

  
  
    
  
    
  

  
    
    
      <h2 id="publications" class="archive__subtitle">publications</h2>
      
    
  
  
    
      





<div class="list__item">
  <article class="archive__item" itemscope itemtype="http://schema.org/CreativeWork">
    

    <h2 class="archive__item-title" itemprop="headline">
      
        <a href="http://localhost:4000/publication/2024-02-17-paper-title-number-4" rel="permalink">Using statistical profiling to decipher hidden chromatin contacts resulting from repeated sequences
</a>
      
    </h2>
    
    

        
          <p>Published in <i>HAL Thesis</i>, 2024 </p>
        

    
    <p class="archive__item-excerpt" itemprop="description"><div style="text-align: justify">
Genomes are not static entities; they evolve and acquire various repeated elements Repetitive genomic elements, crucial yet challenging for NGS analysis, are addressed by Hicberg, an algorithm employing statistical inference to accurately map their reads in diverse sequencing data and, in yeast, reveals novel interactions between retrotransposons and the enigmatic 2-micron plasmid. Applied to yeast, Hicberg further unveils the influence of retrotransposons on chromatin organization and identifies other new genomic interactions, offering a more comprehensive understanding of genome structure and dynamics.
</div>
</p>
    
    
    
      <p>Recommended citation: Sébastien Gradit. Using statistical profiling to decipher hidden chromatin contacts resulting from repeated sequences. Genetics. Sorbonne Université, 2024. English. ⟨NNT : 2024SORUS494⟩. ⟨tel-05010244⟩<br /><a href="https://theses.hal.science/tel-05010244v1/file/140863_GRADIT_2024_archivage.pdf">Download Paper</a></p>
    

  </article>
</div>

    
  
    
      





<div class="list__item">
  <article class="archive__item" itemscope itemtype="http://schema.org/CreativeWork">
    

    <h2 class="archive__item-title" itemprop="headline">
      
        <a href="http://localhost:4000/publications/2025-xx-xx-Hicberg-paper/" rel="permalink">Hicberg:  Reconstruction of genomic signals from repeated elements
</a>
      
    </h2>
    
    

        
          <p>Published in <i>Pending</i>, 2025 </p>
        

    
    <p class="archive__item-excerpt" itemprop="description"><p>This paper introduces Hicberg, a new method to more accurately map the 3D organization of genomes by resolving ambiguities caused by repetitive DNA sequences in Hi-C data, revealing potentially hidden structural features in various organisms.</p>
</p>
    
    
    
      <p><a href=" https://www.biorxiv.org/content/10.1101/2025.06.20.660295v1 ">Download Paper</a></p>
    

  </article>
</div>

    
  

  
    
    
      <h2 id="softwares" class="archive__subtitle">softwares</h2>
      
    
  
  

  
    
    
      <h2 id="talks" class="archive__subtitle">talks</h2>
      
    
  
  
    
      





<div class="list__item">
  <article class="archive__item" itemscope itemtype="http://schema.org/CreativeWork">
    

    <h2 class="archive__item-title" itemprop="headline">
      
        <a href="http://localhost:4000/talks/talk-ismb-2023" rel="permalink">Statistical inference of repeated sequence contacts in Hi-C maps
</a>
      
    </h2>
    
    

        
         <p class="page__date"><strong><i class="fa fa-fw fa-calendar" aria-hidden="true"></i> Published:</strong> <time datetime="2023-07-25T00:00:00+00:00">July 25, 2023</time></p>
        

    
    <p class="archive__item-excerpt" itemprop="description"><p>Increasingly detailed investigations of the spatial organization of genomes reveal that chromosome folding influences or regulates dynamic processes such as transcription, DNA repair and segregation. Hi-C approach is commonly used to characterize genome architecture by quantifying physical contacts’ frequency between pairs of loci through high-throughput sequencing. These sequences cause challenges during the analysis’ alignment step, due to the multiplicity of plausible positions to assign sequencing reads. These unknown parts of the genome architecture, that may contain biological information, remains hidden throughout downstream functional analysis. To overcome these limitations, we have developed <a href="https://github.com/sebgra/hicberg"><em>HiC-BERG</em></a>, a method combining statistical inference with input from DNA polymer behavior characteristics and features of the Hi-C protocol to assign with robust confidence repeated reads in a genome and “fill-in” empty vectors in contact maps. <a href="https://github.com/sebgra/hicberg"><em>HiC-BERG</em></a> is intended to be applicable to different types of organisms. We will present the program and key validation tests, before applying it to unveil hidden parts of the genomes of E.coli, <em>Saccharomyces cerevisiae</em> and <em>Plasmodium falciparum</em>. <a href="https://github.com/sebgra/hicberg"><em>HiC-BERG</em></a> shows that repeated sequences may be involved in singular genomic architectures. Our method can provide an alternative visualization of genomic contacts under a wide variety of biological conditions allowing a more complete view of genome plasticity.</p>

</p>
    
    
    

  </article>
</div>

    
  
    
      





<div class="list__item">
  <article class="archive__item" itemscope itemtype="http://schema.org/CreativeWork">
    

    <h2 class="archive__item-title" itemprop="headline">
      
        <a href="http://localhost:4000/talks/talk-icte-2024" rel="permalink">Hicberg – Hi-C Biological Estimation of Reapeated elements in Genomes
</a>
      
    </h2>
    
    

        
         <p class="page__date"><strong><i class="fa fa-fw fa-calendar" aria-hidden="true"></i> Published:</strong> <time datetime="2024-04-22T00:00:00+00:00">April 22, 2024</time></p>
        

    
    <p class="archive__item-excerpt" itemprop="description"><p>During their evolution, the genomes of micro-organisms can acquire quantities of different repeated elements such as retrotransposons, duplicated genes or tandem repeats. This type of sequence within genomes cannot be processed directly by NGS technologies because they generate short reads that cannot be located unambiguously on reference genomes. This information is filtered out by most current pipelines, leading to incomplete genomic tracks resulting in a significant loss of information on biological functions, processes and genomic structures involving repeated elements. We developed <a href="https://github.com/sebgra/hicberg"><em>Hicberg</em></a>, an algorithm that uses statistical inference and pseudo-random generators to predict the positions of repeated sequence reads from different omics paired-end data (including Hi-C, Mnase-seq or ChIP-seq). After developing the method and calibrating it on a test bench, we explored during this PhD project how it improves genomic data interpretability of various species, starting with microbial ones such as <em>Saccharomyces cerevisiae</em>. Reconstruction of Hi-C and ChIP-seq genomic tracks with <a href="https://github.com/sebgra/hicberg"><em>Hicberg</em></a> revealed how some retrotransposons in this model organism contribute in the positioning of cohesin, a molecular motor involved in the formation of chromatin loops. A new role for retrotransposons sequences as contact points for the elusive yeast 2 micron episomal molecule was also identified. Overall, these results underline the power of the approach to discover new novel molecular relationships, and the interest in applying this tool more widely to larger genomes with greater quantities of repeats. The proposed method can therefore provide an alternative visualization of genomic signals in a wide variety of biological conditions and allow a more comprehensive view of genome organization and plasticity. Importantly, existing datasets can be
revisited using the approach to unveil overlook features.</p>

</p>
    
    
    

  </article>
</div>

    
  
    
      





<div class="list__item">
  <article class="archive__item" itemscope itemtype="http://schema.org/CreativeWork">
    

    <h2 class="archive__item-title" itemprop="headline">
      
        <a href="http://localhost:4000/talks/talk-jobim-2024" rel="permalink">Hicberg : Prediction of omics signals from repeated elements
</a>
      
    </h2>
    
    

        
         <p class="page__date"><strong><i class="fa fa-fw fa-calendar" aria-hidden="true"></i> Published:</strong> <time datetime="2024-06-26T00:00:00+00:00">June 26, 2024</time></p>
        

    
    <p class="archive__item-excerpt" itemprop="description"><p>Repeated genomic elements (like retrotransposons) are poorly mapped by standard NGS pipelines, resulting in a significant loss of information regarding biological functions and genomic structures. To overcome this, we developed <a href="https://github.com/sebgra/hicberg"><em>Hicberg</em></a>, an algorithm that uses statistical inference trained on unambiguous genome regions to predict the positions of reads from these repeated sequences. <em>Hicberg</em> successfully reconstructs complete genomic tracks from various paired-end omics data, including Hi-C and ChIP-seq, significantly improving data completeness and interpretability.</p>

</p>
    
    
    

  </article>
</div>

    
  
    
      





<div class="list__item">
  <article class="archive__item" itemscope itemtype="http://schema.org/CreativeWork">
    

    <h2 class="archive__item-title" itemprop="headline">
      
        <a href="http://localhost:4000/talks/talk-inception-2024" rel="permalink">Using statistical profiling to decipher hidden chromatin contacts resulting from repeated sequences
</a>
      
    </h2>
    
    

        
         <p class="page__date"><strong><i class="fa fa-fw fa-calendar" aria-hidden="true"></i> Published:</strong> <time datetime="2024-11-26T00:00:00+00:00">November 26, 2024</time></p>
        

    
    <p class="archive__item-excerpt" itemprop="description"><p>During their evolution, the genomes of micro-organisms can acquire quantities of different repeated elements such as retrotransposons, duplicated genes or tandem repeats. This type of sequence within genomes cannot be processed directly using high throughput sequencing technologies because they generate short sequences that cannot be located unambiguously on reference genomes. This type of data is filtered by most current standard pipelines, resulting in a significant loss of information on biological functions, processes and genomic structures involving repeated elements. We developed <a href="https://github.com/sebgra/hicberg"><em>Hicberg</em></a>, an algorithm that uses statistical inference and pseudo-random generators to predict the positions of repeated sequence reads from different omics paired-end data (including contact technologies like Hi-C, nucleosomes or proteins positioning technologies like Mnase-seq or ChIP-seq). After developing the method and calibrating it on a test bench, we explored during this PhD project how it improves genomic data interpretability of various species, starting with microbial ones such as <em>Saccharomyces cerevisiae</em>.</p>

</p>
    
    
    

  </article>
</div>

    
  

  
    
    
      <h2 id="teaching" class="archive__subtitle">teaching</h2>
      
    
  
  
    
      





<div class="list__item">
  <article class="archive__item" itemscope itemtype="http://schema.org/CreativeWork">
    

    <h2 class="archive__item-title" itemprop="headline">
      
        <a href="http://localhost:4000/teaching/2015-spring-teaching-1" rel="permalink">Teaching experience 2
</a>
      
    </h2>
    
    

        
          <p> Workshop, <i>University 1, Department</i>, 2015 </p>
        

    
    <p class="archive__item-excerpt" itemprop="description"><p>This is a description of a teaching experience. You can use markdown like any other post.</p>

</p>
    
    
    

  </article>
</div>

    
  

  </div>
</div>


    <div class="page__footer">
      <footer>
        <!-- start custom footer snippets -->
<a href="/sitemap/">Sitemap</a>
<!-- end custom footer snippets -->

        


<div class="page__footer-follow">
  <ul class="social-icons">
    
      <li><strong>Follow:</strong></li>
    
    
      <li><a href="http://github.com/sebgra"><i class="fab fa-github" aria-hidden="true"></i> GitHub</a></li>
    
    
    
    <li><a href="http://localhost:4000/feed.xml"><i class="fa fa-fw fa-rss-square" aria-hidden="true"></i> Feed</a></li>
    
  </ul>
</div>


<div class="page__footer-copyright">
  &copy; 2025 Your Name, Powered by <a href="http://jekyllrb.com" rel="nofollow">Jekyll</a> &amp; <a href="https://github.com/academicpages/academicpages.github.io">AcademicPages</a>, a fork of <a href="https://mademistakes.com/work/minimal-mistakes-jekyll-theme/" rel="nofollow">Minimal Mistakes</a>.<br />
  Site last updated 2025-10-24
</div>

      </footer>
    </div>

    <script src="http://localhost:4000/assets/js/main.min.js"></script>








  </body>
</html>

